package.json

```diff
@@ -33,18 +33,21 @@
   "jest": {
     "modulePaths": [
       "mode_modules",
-      "shared"
-    ]
+      "shared",
+      "server"
+    ] 
   },
   "babel": {
     "presets": [
       "@babel/preset-env",
       "@babel/preset-react"
+    ],
+    "plugins": [
+      "@babel/plugin-transform-runtime"
     ]
   },
   "dependencies": {
     "@babel/core": "^7.16.7",
-    "@babel/plugin-transform-runtime": "^7.16.8",
```

Comment:
Oh, I haven't realize we already have this dependency. I will then leave it in a regular `dependencies` section, because I don't fully understand where else it is used besides testing.

Reply:
It is used in the "build homebrewery" script.

Reply:
Is "build homeberewery" script considered a `dev` thing?

Reply:
Eh.... I suppose we could say that. It is required for the production site to work, since it bundles the js and copies assets over to the proper directories for deployment, but the build script itself isnt actually included in the app.

Reply:
Which reminds me. If you are planning to test serving of static assets like images, etc., I don't think Jest will be able to see them since the resulting "build" directory is not necessarily the same hierarchy as the source code directory.

Reply:
> Eh.... I suppose we could say that. It is required for the production site to work, since it bundles the js and copies assets over to the proper directories for deployment, but the build script itself isnt actually included in the app.

Ok, I don't understand how the deployment works, so I suggest that we leave everything as-is for now, just to make sure that there will be no regressions. We can re-arrange our dependencies later if that would be needed, through a separate PR.

> Which reminds me. If you are planning to test serving of static assets like images, etc., I don't think Jest will be able to see them since the resulting "build" directory is not necessarily the same hierarchy as the source code directory.

I've already encountered this issue with `robots.txt` route. I think we should probably make the root directory for static assets configurable somehow - besides the fact that it will help with testing, it could also be useful in local setups, where static assets directory is set to some user-friendly directory for easier assets management.

Reply:
> I think we should probably make the root directory for static assets configurable somehow

The problem is that it's not just the asset *location*, but the whole folder nesting structure. I.e., we might include specific graphics in a folder somewhere (like the "google Drive" icons) and those get `require()`ed in the appropriate file. But when bundling happens, the whole source folder structure is collapsed into a one-file bundle, which means any assets get similarly reordered/collapsed into a reduced file structure.

Is there any way you can set Jest to run tests off of the bundled app rather than the source code? That would potentially solve several issues:

1) All assets are readily available; we don't have to build a whole secondary structure for testing.
2) Testing is done on code that would actually go into production; a test might pass on the raw source but fail after bundling and we wouldn't catch it.
3) (possibly) Avoid the issues with slow tests?

Reply:
> Is there any way you can set Jest to run tests off of the bundled app (which resides in the /build folder) rather than the raw source code? That would potentially solve several issues:

I'm not entirely sure how this would help, because my understanding is that `build` directory mostly contains _client_ code - the server part says in place and the only benefit here would be in `shared` part, which is according to my understanding could be server-side rendered to speed up the client app.

Additionally, I think that most of HTTP testing will be focused on API part of the app, because that is where "business logic" resides. Regular user-visible routes will likely be tested in less details, i.e. I do not plan to check the exact html returned by server on `/edit` page - I will check for HTTP status code, response headers and at most I will check that brew source text is present on the page. How exactly it is rendered should be tested by corresponding tests for React component, which indeed could benefit from being launched on bundled version of the app.

I think the summary here is that we need more research and discussion about what exactly to test and how? And based on it we should be able to come up with tests implementation and particular technical solutions which would help tests. Like if we don't plan to check the exact html returned by `/edit` page, then we can mock the whole `Homebrewery` React component to simply print it props to reduce amount of unnecessary things performed by test and to simply checks for the essential things like what has been queried from the database and sent back to user.

Reply:
> I've already encountered this issue with robots.txt route. I think we should probably make the root directory for static assets configurable somehow - besides the fact that it will help with testing...

Ah, wait, we define in `app.js` the static assets folder here:
https://github.com/naturalcrit/homebrewery/blob/b2ec0d4a0c2b6eedfeaff495b84aa8f3d1788a20/server/app.js#L65

So is the test is operating under a different `__dirname`? Is there a way to fix that somehow without having to pass in custom paths all the way through the app? I kind of prefer having the deployment paths built off of `__dirname` because it ensures the files are always in the right location relative to the server by default. Making that configurable trickles down to several other places in the code that would need to see that custom path to put files in that new location when building. I think that adds more complexity than I want for something that should be kind of fixed.

> I'm not entirely sure how this would help, because my understanding is that build directory mostly contains client code

Ah yeah. I was not thinking straight on that.

Reply:
> So is the test is operating under a different __dirname? Is there a way to fix that somehow without having to pass in custom paths all the way through the app?

Yes, my understanding is that `jest` launched everything under some other directory. I will dig deeper into documentation to see if we can re-configure `jest` instead of modifying the source code